Spring injection

ioan994 opened the pull request:
Set up injection according to https://www.mkyong.com/webservices/jax-rs/jersey-spring-integration-example/

TanyaBond committed:
spring injection  doesn't work

TanyaBond committed:
Merge branch 'master' of https://github.com/fland/notes-keeper into SpringInjection

# Conflicts:
#	pom.xml

TanyaBond committed:
new Version Spring Context

ioan994 committed:
Set up dependency injection according to https://www.mkyong.com/webservices/jax-rs/jersey-spring-integration-example/

Added tomcat7-maven-plugin to deploy app to comcast with Idea CE

mnovoseltsev commented on src/main/webapp/WEB-INF/applicationContext.xml:
```diff
@@ -2,10 +2,13 @@
-  http://www.springframework.org/schema/context/spring-context.xsd"
->
-    <bean id="greetingService" class="org.noteskeeper.server.GreetingService"/>
+           http://www.springframework.org/schema/beans/spring-beans-2.5.xsd
```
XML-based Spring configs are legacy. Let's use Java configs - https://docs.spring.io/spring/docs/current/spring-framework-reference/htmlsingle/#beans-java. As for mkyong and other 3rd party sites, Spring has pretty awesome own docs and a lot of official examples on github. They  are trying to keep them always up to date. So it is much better to use.

ioan994 replied:
XML-based Spring configs are used in official examples to define beans.
https://jersey.java.net/documentation/latest/spring.html + https://github.com/jersey/jersey/tree/2.25.1/examples/helloworld-spring-webapp

mnovoseltsev replied:
This are Jersey examples, not Spring ones. From Jersey perspective it is Ok to show you XML way to configure, since it still exists in the huge number of legacy applications. There is also annotation driven example - https://github.com/jersey/jersey/tree/2.25.1/examples/helloworld-spring-annotations. But it is pretty crappy. 
XML or Java configs, it is mostly related to Spring. It is not dependent on Jersey. In general I would advise switch to Spring Boot. But maybe it is better to start learning with plain Spring. @fland what do you think?
https://github.com/spring-projects/spring-boot/tree/master/spring-boot-samples/spring-boot-sample-jersey spring boot example.
Also I noted that they use <artifactId>jersey-spring4</artifactId> dependency. Maybe it is more convenient with Spring 4.
Also you may try to browse this repo (without Spring Boot) - https://github.com/peeskillet/underdog-jersey-spring-example/tree/master/src/main, it is not official, but may be handy.
Additionally, as I found, number of people are arguing that spring integration jersey library is not good at all.

fland replied:
Spring XML configs are still widely used (and has its own advantages) and it's a good starting point.
anyway, spring beans 2.5 version schema is extremely outdated. mkyoung example uses spring 3 and relies on dep jersey-spring3 which looks strange with spring 4.
there's no similar dep for spring 4 yet => jersey JAR-RS implementation looks outdated / not actively supported.
I'd strongly recommend to switch to Apache CXF instead of Jersey, as it support Spring better 
Spring Boot doesn't allow to understand all Spring mechanics as it does a lot of thing under the hood like a magic, let's start from 'classic' Spring IoC

mnovoseltsev replied:
XML is widely used in the projects that started before Java Config appeared. As for new projects, I have faced only java configs. Main point is that, Spring boot doesn't support XML, so you would need to mix both approaches, that is not good.  As for just learning matters, it will be good to have some XML configs experience.
BTW, I do not see real benefits in XML Configs that may outweigh Java Config pros.

mnovoseltsev reviewed (dismissed):
Require migration to Spring Java configs. Are we at least trying to use tdd ? Where is the test for endpoint?

mnovoseltsev commented on src/main/java/org/noteskeeper/server/IGreetingService.java:
```diff
@@ -0,0 +1,5 @@
+package org.noteskeeper.server;
+
+public interface IGreetingService {
```
I would prefer to use plural form for interface(GreetingServices) and singular for implementation instead of using such suffixes and prefixes. Just like an idea.

ioan994 replied:
Plural implies collection of items which is not the case with GreetingService. 'Impl' suffix sounds good to me though.

mnovoseltsev replied:
It is not a variable/property name, but name of a type (class/interface). So it is not really associative with plural form. But as I said earlier, it is just a suggestion. If you like *Impl more, let's use it.

ioan994 committed:
Renamed IGreetingService -> GreetingService, GreetingService -> GreetingServiceImpl

fland reviewed (commented):
switch to Apache CXF instead of Jersey

Commit:
Use spring MVC instead of jersey

Commit:
Rename controller class

Commit:
Instruct maven war plugin to ingnore missing web.xml file

We don't need this file since we are using Spring Java configuration

ioan994 commented:
- use Spring MVC instead of Jersey
 - dump any remnants of xml based configs

Commit:
Add endpoint test

Commit:
Fix review comments

ioan994 commented:
Fixed issues that were uncovered during review.

mnovoseltsev commented on src/main/java/org/noteskeeper/server/HelloWorldController.java:
```diff
@@ -0,0 +1,21 @@
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+@Controller
```
For REST services it is more convenient to use meta-annotation '@RestController'. It encapsulates simple '@Controller' and '@RequestBody' annotations. So you may not specifie last one on methods.

mnovoseltsev commented on src/main/java/org/noteskeeper/server/HelloWorldController.java:
```diff
@@ -0,0 +1,21 @@
+import org.springframework.web.bind.annotation.ResponseBody;
+
+@Controller
+@RequestMapping(path="/api")
```
For most annotation there is a more convenient way to put parameter if it is only one and is mapped to 'value' parameter. In that case it is not obligatory to add 'path=', you may just leave '/api'. It might be confusing first time, you just need to understand how does it work.

Commit:
Minor changes

Removed explicit parameter mention in annotations, usage of
@RestController annotation instead of @Controller.
